lib/resty/redis_cluster.lua

```diff
@@ -0,0 +1,402 @@
+    local index = 1 
+    local start = 1 
+    for i = 1, max do
+        local last, delim_last = find(str, delim, start, true)
+        if last == nil or delim_last == nil then
+            break
+        end 
+
+        t[i] = sub(str, start, last - 1)
+        start = delim_last + 1 
+        index = i + 1 
+    end 
+    t[index] = sub(str, start)
+    return t
+end
+
+local mt = { __index = _M }
+
+function _M.new(self, cluster_id, startup_nodes, opt)
+    if clusters[cluster_id] == nil then
+        
+        clusters[cluster_id] = {
+            startup_nodes = startup_nodes,
+            nodes = nil,
+            slots = nil,
+            timeout = (opt and opt.timeout) and opt.timeout or REDIS_CLUSTER_DEFAULT_TIMEOUT,
+            keepalive_size = (opt and opt.keepalive_size) and opt.keepalive_size or REDIS_CLUSTER_DEFAULT_KEEPALIVE_SIZE,
+            keepalive_duration = (opt and opt.keepalive_duration) and opt.keepalive_duration or REDIS_CLUSTER_DEFAULT_KEEPALIVE_DURATION,
+            ttl = (opt and opt.ttl) and opt.ttl or REDIS_CLUSTER_REQUEST_TTL,
+            refresh_table_asap = false,
+            initialized = false
+        }
+    end
+
+    local obj = {
+        cluster_id = cluster_id
+    }
+    return setmetatable(obj, mt)
+end
+
+function _M.initialize(self)
+    local cluster = clusters[self.cluster_id]
+
+    if cluster == nil or cluster.startup_nodes == nil then
+        return nil
+    end
+
+    if cluster.initialized == true and cluster.refresh_table_asap == false then
+        return nil
+    end
+
+    local startup_nodes = cluster["startup_nodes"]
+
+    for i = 1, #startup_nodes do
+        local node = startup_nodes[i]
+        local r = get_redis_link(node[1], node[2], cluster.timeout)
+        local results, err = r:cluster("nodes")
+
+        cluster.nodes = new_tab(500, 0)
+        cluster.slots = new_tab(REDIS_CLUSTER_HASH_SLOTS, 0)
+        
+        if results then
+            local lines = string_split(results, char(10), 1000)
+            for line_index = 1, #lines do
+                local line = lines[line_index]
+                local fields = string_split(line, " ")
+                if #fields > 1 then
+                    local addr_str = fields[2]
+                    local addr = nil
+
+                    if addr_str == ":0" then
+                        addr = { node[1], 
+                                 tonumber(node[2]), 
+                                 node[1] .. ":" .. tostring(node[2]) }
+                    else
+                        local host_port = string_split(addr_str, ":", 2)
+                        addr = { host_port[1], 
+                                 tonumber(host_port[2]), 
+                                 addr_str }
+                    end
+                    cluster.nodes[#(cluster.nodes) + 1] = addr
+                    
+                    local cluster_slots = cluster.slots
+
+                    for slot_index = 9, #fields do
```

Comment:
here，if the slots in one node is 0-1665 5461-7127 10923-12588,  there will be some lose slots when add slot cache